### ISSUE TITLE
File.SetLastAccessTimeUtc on Ubuntu is only up to seconds

### ISSUE THREAD

**Issue by wli3** [dotnet/corefx#31379]
Similar to https://github.com/dotnet/corefx/issues/26024 but for write

This blocks using msbuild incremental build. Especially when using MemoryMappedFile, LastWriteTime will not be set automatically

```c#
using System;
using System.IO;

namespace testLastAccess
{
    class Program
    {
        static void Main(string[] args)
        {
            File.WriteAllText("file_reference", "");
            File.WriteAllText("file_change_lastaccess", "");

            File.SetLastAccessTimeUtc("file_change_lastaccess", DateTime.UtcNow);

            Console.WriteLine(File.GetLastWriteTimeUtc("file_reference").Ticks + " file_reference");
            Console.WriteLine(File.GetLastWriteTimeUtc("file_change_lastaccess").Ticks  + " file_change_lastaccess");
        }
    }
}
```

on Ubuntu
```
wul@willauzrelinux3:~/test/testSetFileLastAccessTime$ dotnet run
636681614066249674 file_reference
636681614060000000 file_change_lastaccess
```

file_change_lastaccess has SetLastAccessTimeUtc after file_reference is created. But end up easier than file_reference. And clearly, the ticks are rounded.

on Windows it is correct
```
λ  dotnet run
636681613739905306 file_reference
636681613739915309 file_change_lastaccess
```

**Comment by danmosemsft** [comment]
@Anipik could you please take a look when you free up?

**Comment by Anipik** [comment]
yep I will take a look at it sometime next week :)

**Comment by danmosemsft** [comment]
I'll set the milestone to 2.1.x for now.

**Comment by wli3** [comment]
It looks like this impact large project solution builds. Currently Ubuntu takes 2 times of time to build incrementally than Windows

**Comment by Anipik** [comment]
yeah it will be an improvement. I will post the numbers when I put the PR to fix this

**Comment by Anipik** [comment]
I am able to make it work using ```utimes```

@wli3 can you confirm that following is the required output ?
```
  Discovering: System.IO.FileSystem.Tests
  Discovered:  System.IO.FileSystem.Tests
  Starting:    System.IO.FileSystem.Tests
  636686696383220724 file_reference
  636686696383221610 file_change_lastaccess
```

**Comment by wli3** [comment]
@Anipik yes, this looks good

**Comment by Anipik** [comment]
@danmosemsft  its fixed in master, do we want it to go for release branch

**Comment by danmosemsft** [comment]
Thanks @anipik you're the file timestamp expert now. Yes, please add the template to this issue, create the port PR, and send mail to the usual alias per the procedure in my email and while back.

**Comment by Anipik** [comment]
Thanks :)  I will do that

**Comment by Anipik** [comment]
### Shiproom template

#### Description
Currently, when we set the LastAccessTime or LastModifiedTime of the files on Unix, the millisecond/micrososecond/nanosecond attribute of the timestamp of the file is always set to zero.

#### Customer Impact
 The customers will be able to set Last Access Time or Last Modified Time in Unix to nanosecond granularity.


#### Regression?
Not a Regression

#### Risk
Low risk because this change is already in master.
We are now using utimensat function to set the timestamp, if the utimensat is not available on the machine we are falling back to utimes

**Comment by vivmishra** [comment]
Approved for 2.1.4

**Comment by Anipik** [comment]
Updated the template with performance numbers

**Comment by karelz** [comment]
Fixed in 3.0 in PR #31522 (setting milestone accordingly & closing).

Note: The 2.1.x port was rejected - see PR #31569.
